File: algorithms/막대기.py
"""
https://www.acmicpc.net/problem/17608
[문제 해결 전략]
1. 입력 받은 값을 정수로 stack에 쌓는다.
2. reversed() 함수를 이용해 오른쪽 끝부터 높이를 비교하며 카운트를 한다.
"""

import sys

N = int(sys.stdin.readline())
stack = []

for _ in range(N):
    stack.append(int(sys.stdin.readline()))

cnt = 0
max_height = 0

for h in reversed(stack):   # 값 자체를 사용해야 함
    if h > max_height:
        max_height = h
        cnt += 1

print(cnt)

File: 막대기.js
/*
 * BOJ 1094 - 막대기
 * https://www.acmicpc.net/problem/1094
 *
 * [Problem & Solvings]
 * - 문제: 길이가 64cm인 막대를 절반씩 자르면서 Xcm 막대를 만들기 위한 최소 막대 개수를 구하라
 * - 조건: 자를 때는 절반으로만 자를 수 있음, 붙이는 것은 자유
 * - 핵심 아이디어: 이진수에서 1의 개수 == 필요한 막대 개수 (2의 제곱수로만 구성됨)
 *
 * [Pseudocode]
 * 1. X를 입력받음
 * 2. X를 이진수로 변환
 * 3. 이진수에서 '1'의 개수를 셈
 * 4. 이 개수를 출력
 *
 * [Time Complexity]
 * - O(log X): 이진수 변환 및 비트 탐색
 * - 입력 X의 최대값이 64이므로 거의 상수 시간
 */

const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

rl.on("line", (line) => {
  const X = parseInt(line.trim());
  const binary = X.toString(2); // 2진수로 변환
  let count = 0;

  for (let bit of binary) {
    if (bit === '1') count++;
  }

  console.log(count);
  rl.close();
});
